### ISSUE TITLE
Useless debugger, wrong source map, shifted breakpoints

### ISSUE THREAD

**Issue by robozb** [expo/expo#24273]
### Summary

The debugger is useless on this way, the breakpoints are shifted:

https://github.com/facebook/metro/assets/8656998/8dea359d-2d94-4255-9116-b2f8d77de24d

```
...
  "dependencies": {
    "expo": "~49.0.8",
    "expo-status-bar": "~1.6.0",
    "react": "18.2.0",
    "react-native": "0.72.4"
  },
  "devDependencies": {
    "@babel/core": "^7.20.0"
  },
...
```

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: Windows 10 10.0.19045        
    Binaries:
      Node: 18.17.1 - C:\Program Files\nodejs\node.EXE
      npm: 9.6.7 - C:\Program Files\nodejs\npm.CMD
    npmPackages:
      expo: ~49.0.8 => 49.0.9
      react: 18.2.0 => 18.2.0
      react-native: 0.72.4 => 0.72.4
    Expo Workflow: managed

### Please specify your device/emulator/simulator platform, model and version
```
Properties
avd.ini.displayname              Pixel 2 API 30
avd.ini.encoding                 UTF-8
AvdId                            Pixel_2_API_30
disk.dataPartition.size          6442450944
fastboot.chosenSnapshotFile
fastboot.forceChosenSnapshotBoot no
fastboot.forceColdBoot           no
fastboot.forceFastBoot           yes
hw.accelerometer                 yes
hw.arc                           false
hw.audioInput                    yes
hw.battery                       yes
hw.camera.back                   virtualscene
hw.camera.front                  emulated
hw.cpu.ncore                     4
hw.device.hash2                  MD5:55acbc835978f326788ed66a5cd4c9a7
hw.device.manufacturer           Google
hw.device.name                   pixel_2
hw.dPad                          no
hw.gps                           yes
hw.gpu.enabled                   yes
hw.gpu.mode                      auto
hw.initialOrientation            Portrait
hw.keyboard                      yes
hw.lcd.density                   420
hw.lcd.height                    1920
hw.lcd.width                     1080
hw.mainKeys                      no
hw.ramSize                       1536
hw.sdCard                        yes
hw.sensors.orientation           yes
hw.sensors.proximity             yes
hw.trackBall                     no
image.androidVersion.api         30
image.sysdir.1                   system-images\android-30\google_apis_playstore\x86\
PlayStore.enabled                true
runtime.network.latency          none
runtime.network.speed            full
showDeviceFrame                  yes
skin.dynamic                     yes
tag.display                      Google Play
tag.id                           google_apis_playstore
vm.heapSize                      228

```

### Error output

Debugger  breakpoints are shifted

### Reproducible demo or steps to reproduce from a blank project

```
import { StatusBar } from 'expo-status-bar';
import { Button, StyleSheet, Text, View } from 'react-native';

export default function App() {

  function a() {
    let a=1
    a=2
    a=3
    a=4
    a=5
    a=6

  }

  return (
    <View style={styles.container}>
      <Text>Open up App.js to start working on your app!</Text>
      <Button title="sajt" onPress={()=>{a()}}></Button>
      <StatusBar style="auto" />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

```

https://github.com/robozb/AwesomeProject

**Comment by robozb** [comment]
It seems the problem is related with expo sdk 49. 

I tried all the things with Expo SDK 48, and the debugger works fine with proper breakpoints!

https://github.com/expo/expo/assets/8656998/65311af8-1e51-4911-b14e-386ea4703c7a

**Comment by robozb** [comment]
Both demo projects:
- https://github.com/robozb/ExpoSdk49-Debugger-Error
- https://github.com/robozb/ExpoSdk48-Debugger-OK

**Comment by brentvatne** [comment]
hi @robozb - this is likely related to this hermes issue: https://github.com/facebook/hermes/issues/1087

**Comment by brentvatne** [comment]
cc @byCedric

**Comment by robozb** [comment]
Thank you @brentvatne !

**Comment by OneHatRepo** [comment]
temporary [workaround here](https://github.com/expo/expo/issues/23990#issuecomment-1682253322).

**Comment by rlgo** [comment]
> temporary [workaround here](https://github.com/expo/expo/issues/23990#issuecomment-1682253322).

This issue occurs on both sdk 49 and sdk 50 preview
And the temporary workaround is not working for me

**Comment by rkostrab** [comment]
Using sdk50 and debugger remains useless.

**Comment by Gtosta96** [comment]
I'm so sad I'm being forced to upgrade from 48 to 49/50. JSC debugger works just fine.

**Comment by OneHatRepo** [comment]
The workaround on 49 works for me, but I know the clock is ticking to where I'll have to upgrade, and I only hope it'll be fixed by the time I have to upgrade.

**Comment by appsgenie** [comment]
same here. upgraded to SDK 49 from 48 and have to use Hermes now but can't debug at all now. Breakpoints in devTools are not pausing and when the debugger statements are pausing the rest of the variables all look off.
are the 2 options to wait on https://github.com/facebook/hermes/issues/1087 or use the [workaround ](https://github.com/expo/expo/issues/23990#issuecomment-1682253322) ? btw @OneHatRepo  for this do we need to rebuild a new dev client?

**Comment by OneHatRepo** [comment]
> btw @OneHatRepo for this do we need to rebuild a new dev client?

I believe you can change the metro-config version without creating a new dev client.

**Comment by OneHatRepo** [comment]
BTW, I'm still using "@expo/metro-config": "^0.7.1" for everything right now for both web and mobile, and haven't encountered any issues with it so far, so I'm able to keep working.

**Comment by appsgenie** [comment]
I only see references to `@expo/metro-config` in the .lock file.. so I guess I never used it? where would I need to add it to so that breakpoints work better with SDK 49 @OneHatRepo ?
And to confirm, this is with `"jsEngine": "hermes"` correct?

**Comment by OneHatRepo** [comment]
@appsgenie the metro-config I was referring to is one of the devDependencies in package.json, and the jsEngine is hermes (this is declared in app.json, though I believe it's now not necessary to declare as it is the default).

**Comment by appsgenie** [comment]
@OneHatRepo, I added that and nothing changed.
restarted the app, invoked debugger with `j` , and breakpoints are not being hit at all :(
even the `debugger;` is not helping. all variables are `undefined`..
using jsEngine=hermes.

```
  "devDependencies": {
    "@babel/core": "^7.19.3",
    "babel-preset-expo": "^9.3.0",
    "@expo/metro-config": "^0.7.1"
  },
```
are you SDK 49 too with custom dev client? You're still able to with in DevTools with breakpoints?

**Comment by OneHatRepo** [comment]
I am able to get it to work perfectly with 49 on a dev client. With 50 in a dev client, it (sometimes) doesn't trigger inline breakpoints, but respects things like "debugger;" and "console.log"

A package.json file from a working install of 49 has this:

  "devDependencies": {
    "@babel/core": "^7.20.0",
    "@expo/metro-config": "^0.7.1"
  },

A package.json file from a (somewhat) working install of 50 has this:

  "devDependencies": {
    "@babel/core": "^7.21.4",
    "@babel/preset-env": "^7.21.4",
    "@babel/runtime": "^7.21.0",
    "@expo/metro-config": "^0.7.1",
    "babel-preset-expo": "^10.0.0"
  },

**Comment by LucasBourgeois** [comment]
Any news ? 
Latest of @expo/metro-config is 0.17.7 .....

In addition of making debug impossible, gzip encoding is still undecoded because of metro-config

**Comment by appsgenie** [comment]
@OneHatRepo unfortunately it still does not show variable values for me during debug. I've used your above mentioned devDependencies for SDK 49 without a new dev build and also after re-building a new dev client and no changes towards good :( no matter what I try, the debugger in Chrome Dev Tools will just not show any valid values for most of the variables.
I tried debugging with VS Code too, but the app crashes after a couple seconds once it pauses on a breakpoint.. so there is really no way to debug an Expo app at this time ...

**Comment by ellouzeskandercs** [comment]
I have investigated this bug recently :

- The bug comes from the serializer, which injects [environment variables into the bundle](https://github.com/expo/expo/blob/359f1da6e77431f18adbaccc10b5110d13266e2d/packages/%40expo/metro-config/src/serializer/environmentVariableSerializerPlugin.ts#L107) : it injects lines of code into the beginning of the bundle and shifts the rest, resulting in shifted source maps, but I cannot understand why 🤷‍♂️
- If you're not using the `EXPO_PUBLIC_*` environment variable injection feature, you can get around it by disabling this feature : `EXPO_NO_CLIENT_ENV_VARS=1 yarn start`
- For the record: in order to inject environment variables, we use a babel plugin instead of the serializer. I think that @EvanBacon fixed that since [their source maps were broken also in release mode](https://github.com/expo/expo/commit/f2bee0b3e60005f90e5fe632b40c41e60175be19) (do you confirm?) → [we don't do it in dev mode, for caching reasons](https://github.com/expo/expo/blob/05d1e0e4b1459c39f00592b24c6e5c04de675fdc/packages/babel-preset-expo/src/common.ts#L96).

A quick fix (and a kind of dirty fix) would be to inject the environment variables into the prelude module using this patch (the patch should be adapted depending on your metro-config package version)
```
diff --git a/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js b/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js
index 0403988..491cd10 100644
--- a/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js
+++ b/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js
@@ -70,13 +70,14 @@ function environmentVariableSerializerPlugin(entryPoint, preModules, graph, opti
     const [firstModule, ...restModules] = preModules;
     // const envCode = `var process=this.process||{};${str}`;
     // process.env
+
+    firstModule.output[0].data.code = firstModule.output[0].data.code + ';' + str;
+
     return [
         entryPoint,
         [
             // First module defines the process.env object.
             firstModule,
-            // Second module modifies the process.env object.
-            getEnvPrelude(str),
             // Now we add the rest
             ...restModules,
         ],
```

**Comment by nicozenf** [comment]
Thank you so much @ellouzeskandercs!
Your `EXPO_NO_CLIENT_ENV_VARS` hack works well, at first glance. We keep an eye on potential side-effets, and keep look forward a real fix.

**Comment by rahmanusta** [comment]
Same issue on SDK 51

**Comment by brycnguyen** [comment]
> I have investigated this bug recently :
> 
> * The bug comes from the serializer, which injects [environment variables into the bundle](https://github.com/expo/expo/blob/359f1da6e77431f18adbaccc10b5110d13266e2d/packages/%40expo/metro-config/src/serializer/environmentVariableSerializerPlugin.ts#L107) : it injects lines of code into the beginning of the bundle and shifts the rest, resulting in shifted source maps, but I cannot understand why 🤷‍♂️
> * If you're not using the `EXPO_PUBLIC_*` environment variable injection feature, you can get around it by disabling this feature : `EXPO_NO_CLIENT_ENV_VARS=1 yarn start`
> * For the record: in order to inject environment variables, we use a babel plugin instead of the serializer in release mode. I think that @EvanBacon fixed that since [the source maps were broken also in release mode](https://github.com/expo/expo/commit/f2bee0b3e60005f90e5fe632b40c41e60175be19) (do you confirm?) → [we don't do it in dev mode, for caching reasons](https://github.com/expo/expo/blob/05d1e0e4b1459c39f00592b24c6e5c04de675fdc/packages/babel-preset-expo/src/common.ts#L96).
> 
> A quick fix (and a kind of dirty fix) would be to move the injection of environment variables code to the prelude module without any line breaks using this patch (the patch should be adapted depending on your metro-config package version)
> 
> ```
> diff --git a/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js b/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js
> index 0403988..491cd10 100644
> --- a/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js
> +++ b/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js
> @@ -70,13 +70,14 @@ function environmentVariableSerializerPlugin(entryPoint, preModules, graph, opti
>      const [firstModule, ...restModules] = preModules;
>      // const envCode = `var process=this.process||{};${str}`;
>      // process.env
> +
> +    firstModule.output[0].data.code = firstModule.output[0].data.code + ';' + str;
> +
>      return [
>          entryPoint,
>          [
>              // First module defines the process.env object.
>              firstModule,
> -            // Second module modifies the process.env object.
> -            getEnvPrelude(str),
>              // Now we add the rest
>              ...restModules,
>          ],
> ```

This worked great for me after patching! I can see the values in my debugger. Thank you so much!

**Comment by robozb** [comment]
> I have investigated this bug recently :
> 
> * The bug comes from the serializer, which injects [environment variables into the bundle](https://github.com/expo/expo/blob/359f1da6e77431f18adbaccc10b5110d13266e2d/packages/%40expo/metro-config/src/serializer/environmentVariableSerializerPlugin.ts#L107) : it injects lines of code into the beginning of the bundle and shifts the rest, resulting in shifted source maps, but I cannot understand why 🤷‍♂️
> * If you're not using the `EXPO_PUBLIC_*` environment variable injection feature, you can get around it by disabling this feature : `EXPO_NO_CLIENT_ENV_VARS=1 yarn start`
> * For the record: in order to inject environment variables, we use a babel plugin instead of the serializer in release mode. I think that @EvanBacon fixed that since [the source maps were broken also in release mode](https://github.com/expo/expo/commit/f2bee0b3e60005f90e5fe632b40c41e60175be19) (do you confirm?) → [we don't do it in dev mode, for caching reasons](https://github.com/expo/expo/blob/05d1e0e4b1459c39f00592b24c6e5c04de675fdc/packages/babel-preset-expo/src/common.ts#L96).
> 
> A quick fix (and a kind of dirty fix) would be to move the injection of environment variables code to the prelude module without any line breaks using this patch (the patch should be adapted depending on your metro-config package version)
> 
> ```
> diff --git a/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js b/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js
> index 0403988..491cd10 100644
> --- a/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js
> +++ b/node_modules/@expo/metro-config/build/serializer/environmentVariableSerializerPlugin.js
> @@ -70,13 +70,14 @@ function environmentVariableSerializerPlugin(entryPoint, preModules, graph, opti
>      const [firstModule, ...restModules] = preModules;
>      // const envCode = `var process=this.process||{};${str}`;
>      // process.env
> +
> +    firstModule.output[0].data.code = firstModule.output[0].data.code + ';' + str;
> +
>      return [
>          entryPoint,
>          [
>              // First module defines the process.env object.
>              firstModule,
> -            // Second module modifies the process.env object.
> -            getEnvPrelude(str),
>              // Now we add the rest
>              ...restModules,
>          ],
> ```

Amazing, thank you so much!

**Comment by HJK181** [comment]
Facing the same issue, are there any updates when a fix can be expected?

**Comment by navignaw** [comment]
I encountered the same issue on SDK 50, but @ellouzeskandercs's [solution](https://github.com/expo/expo/issues/24273#issuecomment-2132297993) worked like a charm! 

It looks like this was fixed in #29463 which made it into `@expo/cli` [0.18.17](https://github.com/expo/expo/blob/main/packages/%40expo/cli/CHANGELOG.md#01817---2024-06-10) - @EvanBacon can you confirm?

Presumably this fix won't get backported to Expo SDK 50?

**Comment by OneHatRepo** [comment]
I've been using the EXPO_NO_CLIENT_ENV_VARS hack, so can't tell if the issue still exists. Does anyone know?

**Comment by OfirD1** [comment]
Still happens in SDK 51.0.38 (and the `XPO_NO_CLIENT_ENV_VARS` hack works as well). 

I'll also add that to make it work through the package.json `start` script, then* it can be changed from:
`"start": "expo start"`
to:
`"start": "node -r dotenv/config node_modules/expo/AppEntry.js"`

\* assuming you added the required `.env` file in the root folder, and installed `npm i dotenv`

**Comment by julienvedebe** [comment]
Hi still happens in SDK 52 or it is another one but unfortunetly the workarround is not ok for me.

**Comment by brentvatne** [comment]
@julienvedebe - if you're still having this issue with sdk 52 can you please create a new issue with a minimal reproducible example and steps to follow to reproduce it? thank you!
